src/puppetlabs/puppetdb/threadpool.clj

```diff
@@ -0,0 +1,130 @@
+            ConnectionFactory Connection Queue Message]
+           [java.util.concurrent Semaphore ThreadPoolExecutor TimeUnit SynchronousQueue
+            RejectedExecutionException ExecutorService]
+           [org.apache.commons.lang3.concurrent BasicThreadFactory BasicThreadFactory$Builder])
+  (:require [clojure.core.async :as async]
+            [clojure.tools.logging :as log]
+            [puppetlabs.i18n.core :as i18n]
+            [slingshot.slingshot :refer [throw+]]))
+
+(def logging-exception-handler
+  "Exception handler that ensures any uncaught exception that occurs
+  on this thread is logged"
+  (reify Thread$UncaughtExceptionHandler
+    (uncaughtException [_ thread throwable]
+      (log/error throwable
+                 (i18n/trs "Error processing command")))))
+
+(defn thread-factory
+  "Creates a command thread factory, wrapping the
+  `pool-thread-factory`. Using this thread factory ensures that we
+  have distinguishable names for threads in this pool and that all
+  uncaught exceptions are logged as errors. Without this uncaught
+  exceptions get the default which is standard error"
+  [threadpool-name-pattern pool-thread-factory]
+  (-> (BasicThreadFactory$Builder.)
+      (.wrappedFactory pool-thread-factory)
+      (.namingPattern threadpool-name-pattern)
+      (.uncaughtExceptionHandler logging-exception-handler)
+      (.daemon true)
+      .build))
+
+(defn shutdown
+  "Shuts down the gated threadpool, ensuring in-flight work is
+  complete before tearing it down. Waits up to `shutdown-timeout` (in
+  milliseconds) for that work to complete before forcibly shutting
+  down the threadpool that work to complete"
+  [{:keys [^Semaphore semaphore ^ExecutorService threadpool shutdown-timeout]}]
+
+  ;; Prevent new work from being accepted
+  (.drainPermits semaphore)
+
+  ;; Shutdown the threadpool, allowing in-flight work to finish
+  (.shutdown threadpool)
+
+  ;; This will block, waiting for the threadpool to shutdown
+  (when-not (.awaitTermination threadpool shutdown-timeout java.util.concurrent.TimeUnit/MILLISECONDS)
+
+    (log/warn
+     (i18n/trs "Threadpool not stopped after {0} milliseconds, forcibly shutting it down" shutdown-timeout))
+
+    ;; This will force the shutdown of the threadpool and will
+    ;; not allow current threads to finish
+    (.shutdownNow threadpool)
+    (log/warn
+     (i18n/trs "Threadpool forcibly shutdown"))))
+
+(defrecord GatedThreadpool [^Semaphore semaphore ^ExecutorService threadpool shutdown-timeout]
+  java.io.Closeable
+  (close [this]
+    (shutdown this)))
+
+(defn create-threadpool
+  "Creates an unbounded threadpool with the intent that access to the
+  threadpool is bounded by the semaphore. Implicitly the threadpool is
+  bounded by `size`, but since the semaphore is handling that aspect,
+  it's more efficient to use an unbounded pool and not duplicate the
+  constraint in both the semaphore and the threadpool"
+  [size name-pattern shutdown-timeout-in-ms]
```

Comment:
It's included in the semaphore below. If size wasn't used, that would be really bad :-)